### PR TITLE
Project Completed: Advanced Building Blocks - Enumerables

### PR DESCRIPTION
**Pull request by ivanid22**
The assignment for the project was to do my own implementation on a number of methods for the Enumerable class:

- my_each
- my_each_with_index
- my_select
- my_all
- my_any
- my_none
- my_count
- my_map
- my_inject

Please review this pull request, I'm eager to get your feedback.
Thanks!

### CODE REVIEW

File: main.rb [hunk 1]
```diff
@@ -0,0 +1,172 @@
+module Enumerable
+  def my_each
+    pos = 0
+    loop do
+      break if pos >= length
+
+      yield self[pos]
+      pos += 1
+    end
+  end
```
Comment (stratospherique) [inline main.rb]: * [x] `my_each` is identical and returns the same thing as ruby's [`each`](https://ruby-doc.org/core-2.6.4/Array.html#method-i-each). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

* [x] my_each returns an Enumerator if no block is given
  ```ruby
    array = [1,2,5,8,7,98,4]
    p array.my_each.class # not Enumerator
  ```

File: main.rb [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+  def my_each_with_index
+    pos = 0
+    loop do
+      break if pos >= length
+
+      yield self[pos], pos
+      pos += 1
+    end
+  end
```
Comment (stratospherique) [inline main.rb]: * [x] `my_each_with_index` is identical and returns the same thing as ruby's [`each_with_index`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-each_with_index). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each_with_index.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

* [x] my_each_with_index returns an enumerator if no block is given
  ```ruby
    array = [1,2,5,8,7,98,4]
    p array.my_each_with_index.class # not Enumerator
  ```

File: main.rb [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+
+  def my_select
+    return_arr = []
+    my_each do |element|
+      return_arr.push(element) if yield element
+    end
+    return_arr
+  end
```
Comment (stratospherique) [inline main.rb]: * [x] `my_select` is identical and returns the same thing as ruby's [`select`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-select). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_select.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

* [x] my_select returns an enumerator if no block is given
  ```
    array = [1,2,5,8,7,98,4]
    p array.my_select.class # not Enumerator
  ```

File: main.rb [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+    return_val
+  end
+
+  def my_none
+    return_val = true
+    my_each do |element|
+      return_val = !(yield element)
+    end
+    return_val
+  end
```
Comment (stratospherique) [inline main.rb]: The issues with **my_none?**, **my_any?** and **my_all?** are similar. the methods should be able to: 

* [x] Handle the cases where no block is given (Hint: block_given?)
* [x] Accept primitives, classes (i.e Integer, String, etc...) and Regex Expressions as arguments.

File: main.rb [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+
+  def my_count(arg = nil)
+    count_val = 0
+    my_each do |element|
+      count_val += 1 if (block_given? && yield(element)) || (element == arg)
+    end
+    count_val
+  end
```
Comment (stratospherique) [inline main.rb]: * [x] `my_count` is identical and returns the same thing as ruby's [`count`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-count). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_count.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

* [x] my_count returns the number of items in enum through enumeration
  ```ruby
    array = [1,2,5,4,7]
    p array.my_count == array.count # false
  ```

File: main.rb [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+      if proc_arg
+        return_arr.push(proc_arg.call(element))
+      elsif block_given?
+        return_arr.push(yield(element))
+      else
+        return_arr.push(index)
+      end
+    end
+    return_arr
+  end
```
Comment (stratospherique) [inline main.rb]: * [x] `my_map` is identical and returns the same thing as ruby's [`map`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_map.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

* [x] my_map returns an Enumerator if no block is given
   ```ruby
    array = [1,5,8,7,9,5,5,55,5]
    p array.my_map.class # not an Enumerator
  ```

File: main.rb [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+    memo = args[0].is_a?(Symbol) ? self[0] : args[0] || self[0]
+    symbol = identify_symbol(*args)
+    pos = settle_start_position(*args)
+    while pos < length
+      memo = block_given? ? yield(memo, self[pos]) : memo.send(symbol, self[pos])
+      pos += 1
+    end
+    memo
+  end
+end
```
Comment (stratospherique) [inline main.rb]: * [x] `my_inject` is identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

* [x] my_inject combines all elements of enum by applying a binary operation, specified by a block
  ```ruby
    range = Range.new(5, 50)
    p range.my_inject(4) { |prod, n| prod * n } # returns this error
    #NameError:
        undefined local variable or method `length' for 5..50:Range
      # ./main.rb:80:in `my_inject'
  ```

* [x] my_inject when a symbol is specified combines each element of the collection by applying the symbol as a named method
  ```ruby
    range = Range.new(5, 50)
    p range.my_inject(2, :*) # returns this error
    #  NameError:
        undefined local variable or method `length' for 5..50:Range
      # ./main.rb:80:in `my_inject'
  ```

**Review by stratospherique — CHANGES_REQUESTED** [review]
### Great start @ivanid22 :clap: :clap: However, some changes need to be implemented in order for us to approve your project. Check :arrow_down: 

- **[REQUIRED]** You should name your methods `my_all?`, `my_any?` and `my_none?` instead of their current namings.

- **[REQUIRED]** Delete all the debugging/testing code in `main.rb`  file

- **[REQUIRED]** Check the code comments below, they should help you refactor your methods :gear: 

### That's it :relieved: When finished, kindly resubmit another code review request :pray: 

Happy coding :v: 
**_Ahmed Mahfoudh_**

**Comment by ivanid22** [comment]
Thank you for your feedback Ahmed. I'll be pushing my corrections asap.

**Review by fatymahmed — CHANGES_REQUESTED** [review]
## Hi @ivanid22,
Great job on making some of the changes requested by the previous TSE 👏👏, however, there are a few that you haven't made.

Check my comments below for a detailed review on your methods. I also suggested you use [this](https://ruby-doc.org/core-2.6.5/Enumerable.html) document as a guide to test your methods, and see all the different ways and arguments for each method.

### Status: Changes required 🛑

File: main.rb [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+  end
+
+  def my_none?(arg = nil)
+    return_val = true
+    my_each do |element|
+      return_val = !((block_given? && yield(element)) || (arg == element))
+      return false unless return_val
+    end
+    return_val
+  end
```
Comment (fatymahmed) [inline main.rb]: - For `my_all?`, `my_any?`, `my_none?` handle the following cases
1. when no block or argument is given,
- For the built-in method
`[1, true, 'hi', [ ] ].all?` returns `true`

- Your methods
`[1, true, 'hi', [ ] ].my_all?` returns `false`

2. when the argument is a REGEX or a CLASS.

- For the built-in method
`%w[ant bear cat].all?(/a/) => true`
`[1, 2i, 3.14].all?(Numeric) => true`

- Your methods 
`%w[ant bear cat].my_all?(/a/) => false`
`[1, 2i, 3.14].my_all?(Numeric) => false`

**Comment by ivanid22** [comment]
Thank you Fatima, uploading corrections in a moment.

File: main.rb [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+
+  def my_none?(arg = nil)
+    return_val = true
+    my_each do |element|
+      return_val = !((block_given? && yield(element)) || (arg === element))
+      return_val = true?(element) if !block_given? && !arg
+      return false if return_val
+    end
+    true
+  end
```
Comment (suretrust) [inline main.rb]: * [x] should have working `my_none?` identical and returns the same thing as ruby's [`none?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_none.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
```ruby
[1,2,3,4,5,6,7,8,9].my_none?{ |num| num > 9 } != [1,2,3,4,5,6,7,8,9].none?{ |num| num > 9 }

[1,2,3,4,5].my_none?(String) != [1,2,3,4,5].none?(String)

['saheed','oladele','suretrust'].my_none?(/z/) != ['saheed','oladele','suretrust'].none?(/z/)

[1,2,3,4,5].my_none?(5) != [1,2,3,4,5].none?(5)
```

**Review by suretrust — CHANGES_REQUESTED** [review]
Hi @ivanid22,

### Great job implementing the suggested changes! 🙌 The `my_none` method still needs some touches. 👇 

**Kindly submit for another code review request once these issues have been fixed!** _Happy coding!_ 🎉

File: main.rb (hunk 9 above)
Comment (ivanid22) [reply main.rb]: Thanks for your feedback Saheed. That was an oversight on my part. I'll be committing a solution shortly

**Comment by shshamim63** [comment]
## Status:  :tada: Approved :tada: 
___
Hi @ivanid22 :wave: ,
congratulation :fireworks: . You have successfully completed this project. I can see a descriptive readme. All the methods function like the original one. Code looks neat and clean. Awesome job :clap: . Nothing is blocking me to approve this project. 
___
##### This :+1: is from me, You can go ahead. 
___
## :herb: Happy Coding :herb: 
### Shakhawat Hossain
[Linkedin](https://www.linkedin.com/in/shakhawathossainshamim/)

**Pull request merged**